### ISSUE TITLE
implement `Default` for arrays

### ISSUE THREAD

**Issue by vks** [rust-lang/rust#25169]
Something like

```rust
impl<T: Default> for [T; 2] {                                                       
    fn default() -> [T; 2] {                                                        
        [Default::default(), Default::default()]                                    
    }                                                                               
}
```

for N=0..33 should be added to `src/libcore/default.rs`. The question is how to do that with a macro.

**Comment by kballard** [comment]
What's special about N=0..33?

**Comment by vks** [comment]
Due to lack of generics over integers, we cannot implement it for all N. Typically traits on arrays in the standard library are implemented for N=0..33, see for example `Eq`.

**Comment by nham** [comment]
I looked at this a bit today and I'm stuck. Here's one way to do it, but it requires an extraneous `Copy` bound on `T` (inside libcore/array.rs macro):

```
            impl<T: Copy + Default> Default for [T; $N] {
                #[inline]
                fn default() -> [T; $N] {
                   [$(Default::default() ,)+]
                   [Default::default(); $N]
                }                                                                               
            }
```

Not sure if there's any way to achieve the same effect as array initialization without requiring `Copy`.

**Comment by vks** [comment]
You could write `[Defaul::default(), Default::default(), ...]` instead.

**Comment by shepmaster** [comment]
I think you *have* to write it out as `[Default::default(), Default::default(), ...]`. If the `Default` implementation actually calls something stateful, like `rand`, you presumably would want there to be *different* random values for each item.

**Comment by apasel422** [comment]
This was implemented in #27825 and can be closed.
